sklearn/utils/estimator_checks.py

```diff
@@ -804,6 +805,22 @@ def check_estimators_overwrite_params(name, Estimator):
                      % (name, k, v, new_params[k]))
 
 
+def check_no_fit_attributes_set_in_init(name, Estimator, allowed_properties):
+    """Check that Estimator.__init__ doesn't set trailing-_ attributes.
+
+    allowed_properties is a set of attribute names that are allowed to be
+    on the Estimator class as properties.
+    """
+    estimator = Estimator()
+    for attr in dir(estimator):
+        if attr.endswith("_") and not attr.startswith("_"):
+            assert_true(hasattr(Estimator, attr),
+                        "%s set by __init__ in %s" % (attr, name))
+            assert_true(isinstance(getattr(Estimator, attr), property),
```

Comment:
We do: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/metaestimators.py#L13